### ISSUE TITLE
dev:module:rewrite:conflicts Not Found Error

### ISSUE THREAD

**Issue by sprankhub** [netz98/n98-magerun#856]
If a rewrite is defined on a class which is not available in the Magento installation, `dev:module:rewrite:conflicts` will result in a fatal error:

    PHP Fatal error:  Class 'Some_Class' not found in ...

This is the case when an extension e.g. rewrites an enterprise class, which is not available in the community edition. Example:

    <enterprise_cms>
        <rewrite>
            <config>Some_Extension_Model_Rewrite_Enterprise_CmsConfig</config>
        </rewrite>
    </enterprise_cms>

And the PHP class:

    <?php

    class Some_Extension_Model_Rewrite_Enterprise_CmsConfig extends Enterprise_Cms_Model_Config
    {
        // ...
    }

Magento normally ignores the `enterprise_cms` rewrite node, because it is not available. magerun should do that, too.

**Comment by tkn98** [comment]
So should non-existent rewrite classes be reported or not? What's your suggestion?

Technically I'd say this is worth a warning in verbose mode.

Regardless of the severity, the fatal error must go, that qualifies as a bug.

**Comment by tkn98** [comment]
@sprankhub: Can you provide a location in the code where the fatal error happens? Or a backtrace (e.g. run with xdebug enabled).

**Comment by sprankhub** [comment]
Warning in verbose mode seems okay for me. For normal (non-verbose) usage, I would ignore it. Otherwise, there are too many useless warnings. I guess it is quite common to handle enterprise classes, which are non-existent in CE.

    PHP Stack trace:
    PHP   1. {main}() /usr/local/bin/n98-magerun.phar:0
    PHP   2. N98\Magento\Application->run() /usr/local/bin/n98-magerun.phar:8
    PHP   3. Symfony\Component\Console\Application->run() phar:///usr/local/bin/n98-magerun.phar/src/N98/Magento/Application.php:669
    PHP   4. N98\Magento\Application->doRun() phar:///usr/local/bin/n98-magerun.phar/vendor/symfony/console/Application.php:126
    PHP   5. Symfony\Component\Console\Application->doRun() phar:///usr/local/bin/n98-magerun.phar/src/N98/Magento/Application.php:603
    PHP   6. Symfony\Component\Console\Application->doRunCommand() phar:///usr/local/bin/n98-magerun.phar/vendor/symfony/console/Application.php:195
    PHP   7. N98\Magento\Command\AbstractMagentoCommand->run() phar:///usr/local/bin/n98-magerun.phar/vendor/symfony/console/Application.php:886
    PHP   8. Symfony\Component\Console\Command\Command->run() phar:///usr/local/bin/n98-magerun.phar/src/N98/Magento/Command/AbstractMagentoCommand.php:440
    PHP   9. N98\Magento\Command\Developer\Module\Rewrite\ConflictsCommand->execute() phar:///usr/local/bin/n98-magerun.phar/vendor/symfony/console/Command/Command.php:259
    PHP  10. N98\Magento\Command\Developer\Module\Rewrite\ConflictsCommand->_isInheritanceConflict() phar:///usr/local/bin/n98-magerun.phar/src/N98/Magento/Command/Developer/Module/Rewrite/ConflictsCommand.php:62
    PHP  11. class_exists() phar:///usr/local/bin/n98-magerun.phar/src/N98/Magento/Command/Developer/Module/Rewrite/ConflictsCommand.php:154
    PHP  12. spl_autoload_call() phar:///usr/local/bin/n98-magerun.phar/src/N98/Magento/Command/Developer/Module/Rewrite/ConflictsCommand.php:154
    PHP  13. Varien_Autoload->autoload() phar:///usr/local/bin/n98-magerun.phar/src/N98/Magento/Command/Developer/Module/Rewrite/ConflictsCommand.php:154
    PHP  14. include() /var/www/store/lib/Varien/Autoload.php:94
    PHP  15. spl_autoload_call() /var/www/store/app/code/community/Some/Extension/Model/Rewrite/Enterprise/CmsConfig.php:11
    PHP  16. Varien_Autoload->autoload() /var/www/store/app/code/community/Some/Extension/Model/Rewrite/Enterprise/CmsConfig.php:11
    PHP  17. include() /var/www/store/lib/Varien/Autoload.php:94
    PHP  18. spl_autoload_call() /var/www/store/app/code/community/Some/Extension/Model/Rewrite/Enterprise/CmsConfigResolver.php:12
    PHP  19. Varien_Autoload->autoload() /var/www/store/app/code/community/Some/Extension/Model/Rewrite/Enterprise/CmsConfigResolver.php:12
    PHP  20. include() /var/www/store/lib/Varien/Autoload.php:94

**Comment by tkn98** [comment]
Puh, I'm wondering how this is handled within Magento as this Varien autoloader triggers the fatal error. And we need to trigger autoloading otherwise I don't see how we should resolve classname existence.

Any idea?

**Comment by sprankhub** [comment]
No ad-hoc idea. I will keep this in my inbox and I will try to check that some time, but my time is quite limited currently :-/

**Comment by tkn98** [comment]
Sure, mine as well ;). Fatal errors of this kind are PITA material prior PHP 7. I take this with me as well maybe tomorrow morning I have an idea.

**Comment by tkn98** [comment]
@sprankhub: Can you crate a little PHP script and use `Mage::getModel()` with the string-identifier of the rewritten class and see what happens? Does it fatal error as well?

**Comment by sprankhub** [comment]
Yes, a direct call to `Mage::getModel()` with the string identifier of the rewritten class reveals the same error, which is logical. The issue is that there is no call to `Mage::getModel()` with this identifier in the code base. There is only the defined rewrite (and another rewrite definition it conflicts with).

I think  the issue is that you [here](https://github.com/netz98/n98-magerun/blob/master/src/N98/Magento/Command/Developer/Module/Rewrite/ConflictsCommand.php#L154) check the classes with `get_class`, although the classes may not be relevant since the model string may not be present in the code base. My suggestion would be to filter rewrites, which are not relevant to the current system [here](https://github.com/netz98/n98-magerun/blob/master/src/N98/Magento/Command/Developer/Module/Rewrite/AbstractRewriteCommand.php#L21). So if a rewrite is defined on a model string `enterprise_cms`, but `enterprise_cms` is not present as a model definition, it will not be included in the rewrite commands at all.

**Comment by tkn98** [comment]
This is not necessarily logical. I can configure rewrite that use classes which extend from non-existent classes with without seeing any fatal error. I just checked that already this morning.

I suspect there is an underlying issue in the file that is being included which is causing the fatal error. Can you give a simple include on that file alone a try? And can you report which PHP version you're using? I tested with current old stable PHP 5.6.

**Comment by sprankhub** [comment]
If I include the file with `require_once`, I also get a fatal error. And I think this is default behaviour - how should PHP construct a class without a working inheritance chain? My PHP version is 5.5.35-1+donate.sury.org~trusty+4.

I will send you my test files via mail.

What do you think about my suggestion?

**Comment by tkn98** [comment]
Well if you do require, then it's a fatal error. Because there is a requirement with require, while with include it's a warning.

However Magento's Varien lib autoloader is using *`include`* and not *`require`*. So please test with include.

And take care here that this is a special scenario: The *autoloader* tries something to resolve a class. This *can* fail (non fatally) and then class_exists() returns just false (because even auto-loading was not able to find a specific class's definition, hence the class does not exist).

**Comment by sprankhub** [comment]
With `include_once`, I still get a fatal error.

**Comment by robbieaverill** [comment]
It's worth noting that we now have a "does class exist" check in the `dev:class:lookup` command which might be useful here as well.

**Comment by sprankhub** [comment]
I just checked the `856-fatal-error-on-parent-class-inexistence` and it solves the issue for me. Thanks a lot @ktomk!

**Comment by tkn98** [comment]
@robbieaverill: Do you get fatal errors with that command? If so, it would be a good candidate to continue on. Can you check and create a new issue if that is the case?

**Comment by robbieaverill** [comment]
Will check tomorrow and report back.

>

-- 
*Robbie Averill*
LA PETITE MANOUCHE - NZ Gypsy Jazz Guitar Duo
Phone: +64 27 321 4758
contact@lapetitemanouche.com
www.lapetitemanouche.com

**Comment by robbieaverill** [comment]
Running version 1.97.22 this issue doesn't occur for me, it simply doesn't report the missing class as a conflict.

**Comment by ktomk** [comment]
@robbieaverill: Interesting and thanks for the check. Sounds good to me then, just in case you run into trouble, open a new issue, we can reference this one then.
